src/docfx/lib/process/InterProcessReaderWriterLock.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+
+namespace Microsoft.Docs.Build
+{
+    public struct InterProcessReaderWriterLock : IDisposable
+    {
+        private readonly FileStream _fileStream;
+        private readonly string _filePath;
+
+        private InterProcessReaderWriterLock(FileStream fileStream, string filePath)
+        {
+            _fileStream = fileStream;
+            _filePath = filePath;
+        }
+
+        public static InterProcessReaderWriterLock CreateReaderLock(string lockName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(lockName));
+
+            var filePath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(lockName));
+            var fileLock = WaitFile(lockName, filePath, FileAccess.Read, FileShare.Read);
+
+            return new InterProcessReaderWriterLock(fileLock, filePath);
+        }
+
+        public static InterProcessReaderWriterLock CreateWriterLock(string lockName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(lockName));
+
+            var filePath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(lockName));
```

Comment:
There is already a hash operation in `GlobalMutex`, do we need a nested hash?

Reply:
![image](https://user-images.githubusercontent.com/19990166/65029009-98394c80-d96f-11e9-8b33-10236f34fb74.png)

Reply:
lock name can be remote url

Reply:
`HashUtility.GetMd5Hash(lockName)` can be shared though